src/sentry/replays/consumers/recording/process_recording.py

```diff
@@ -0,0 +1,191 @@
+from __future__ import annotations
+
+import concurrent.futures
+import logging
+import time
+from collections import deque
+from concurrent.futures import ALL_COMPLETED, Future, wait
+from io import BytesIO
+from typing import Callable, Deque, Mapping, MutableMapping, NamedTuple, Optional, cast
+
+import msgpack
+import sentry_sdk
+from arroyo import Partition
+from arroyo.backends.kafka.consumer import KafkaPayload
+from arroyo.processing.strategies.abstract import ProcessingStrategy
+from arroyo.types import Message, Position
+from django.conf import settings
+
+from sentry.attachments import MissingAttachmentChunks, attachment_cache
+from sentry.models import File
+from sentry.replays.consumers.recording.types import RecordingChunkMessage, RecordingMessage
+from sentry.replays.models import ReplayRecordingSegment
+from sentry.utils import json
+
+logger = logging.getLogger("sentry.replays")
+
+CACHE_TIMEOUT = 3600
+
+
+class ReplayRecordingMessageFuture(NamedTuple):
+    """
+    Map a submitted message to a Future returned by the Producer.
+    This is useful for being able to commit the latest offset back
+    to the original consumer.
+    """
+
+    message: Message[KafkaPayload]
+    future: Future[bool | None]
+
+
+class ProcessRecordingStrategy(ProcessingStrategy[KafkaPayload]):
+    def __init__(
+        self,
+        commit: Callable[[Mapping[Partition, Position]], None],
+    ) -> None:
+        self.__closed = False
+        self.__futures: Deque[ReplayRecordingMessageFuture] = deque()
+        self.__threadpool = concurrent.futures.ThreadPoolExecutor()
+        self.__commit = commit
+
+    def poll(self) -> None:
+        self._commit_and_prune_futures()
+
+    def _process_chunk(
+        self, message_dict: RecordingChunkMessage, message: Message[KafkaPayload]
+    ) -> None:
+
+        # can't make this a future as we need to guarentee this happens before
+        # the final kafka message
+        # TODO: determine if this is acceptable at scale
+        id = message_dict["id"]
```

Comment:
as @cmanallen implemented here, https://github.com/getsentry/replay-backend/pull/38/files, we can make this a future by creating a dict with each recording segment chunk's futures, and wait on the final message for those futures to be completed. is that the ideal solution?